src/Evented.ts

```diff
@@ -170,26 +109,17 @@ export class Evented extends Destroyable implements Evented {
 	 *
 	 * @return {any}
 	 */
-	on: BaseEventedEvents = function (this: Evented, ...args: any[]) {
-		if (args.length === 2) {
-			const [ type, listeners ] = <[ string, EventedListenerOrArray<any, EventTargettedObject<any>>]> args;
-			if (Array.isArray(listeners)) {
-				const handles = listeners.map((listener) => aspectOn(this.listenersMap, type, resolveListener(listener)));
-				return handlesArraytoHandle(handles);
-			}
-			else {
-				return aspectOn(this.listenersMap, type, resolveListener(listeners));
-			}
-		}
-		else if (args.length === 1) {
-			const [ listenerMapArg ] = <[EventedListenersMap<any>]> args;
-			const handles = Object.keys(listenerMapArg).map((type) => this.on(type, listenerMapArg[type]));
+	on<K extends keyof M>(type: K, listener: EventedCallbackOrArray<K, M[K]>): Handle;
+	on(type: T, listener: EventedCallbackOrArray<T, O>): Handle;
+	on(type: any, listener: EventedCallbackOrArray<any, any>): Handle {
```

Comment:
Assume this can't be `T | K`?

Reply:
Correct. `K` and `T` need to be specifically associated with the correct `EventedCallbackOrArray`. Using a union type means that `T` could be associated with `EventedCallbackOrArray<K, M[K]>` which we don't want.